### PR TITLE
LSTM SymInt-aware changes & meta registration (non-cuDNN CUDA)

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90980
* #90944
* __->__ #90701
* #90702
* #90714
* #90697

Adds meta registrations for cuDNN and vanilla CUDA ops underneath `lstm()` and makes the logic SymInt-aware.
TODO:
* cuDNN side does some [nasty stuff](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/RNN.cpp#L1567) with buffers; this needs larger redesign to figure out
* Indicate that AOT Autograd can be used when an LSTM is present (remove the check for this once it's fully supported)

### CODE REVIEW

**Comment by ezyang** [comment]
btw, the thnn stuff is independently useful, and I think unblocks some models

**Review by ezyang — COMMENTED** [review]
need tests for the new metas

**Comment by jbschlosser** [comment]
> need tests for the new metas

Working on tests now! Since there's no OpInfo for LSTM, we can either do a one-off test or take advantage of the existing ModuleInfo entries to generate a new set of tests.

> btw, the thnn stuff is independently useful, and I think unblocks some models

I'll split this into separate PRs for thnn / cuDNN

**Review by ezyang — APPROVED** [review]

File: test/test_fake_tensor.py [hunk 1]
```diff
@@ -370,6 +371,36 @@ def fn(
...
+
+                h_0 = torch.randn((num_layers * D, N, H_out), device='cuda')
+                c_0 = torch.randn((num_layers * D, N, hidden_size), device='cuda')
+                inp = torch.randn((L, N, H_in), device='cuda')
+                (output, (h_n, c_n)) = lstm(inp, (h_0, c_0))
+                output.sum().backward()
+
+                self.assertEqual(output.shape, (L, N, D * H_out))
+                self.assertEqual(h_n.shape, (D * num_layers, N, H_out))
+                self.assertEqual(c_n.shape, (D * num_layers, N, hidden_size))
```
Comment (ezyang) [inline test/test_fake_tensor.py:402]: One downside for this test is it doesn't exercise symbolic shapes. But I guess this doesn't actually work yet?

Comment (jbschlosser) [reply test/test_fake_tensor.py:401]: That's true - the test should ideally be at the AOTAutograd level to exercise symbolic shapes, but that doesn't work yet due to #90500.

The test here just covers the meta registration for now. Tests that will exercise symbolic shapes are added in #90980 (currently with xfails for the RNNs).

**Pull request closed without merging**